ipamd/ipamd.go

```diff
@@ -49,6 +50,10 @@ type IPAMContext struct {
 
 	currentMaxAddrsPerENI int
 	maxAddrsPerENI        int
+	// maxENI indicate the maximum number of ENIs can be attached to the instance
+	// It is initialized to 0 and it is set to current number of ENIs attached
+	// when ipamD receives AttachmentLimitExceeded error
+	maxENI int
```

Comment:
For now this should be ok, but in near future we should look into implementing something in a way that even if ipamD restarts we don't invoke AttachENI api to know if max number of ENIs are attached or not.

Reply:
We should create an issue or add TODO to visit this logic later.

Reply:
Or we should probably use the hardcoded map instead of this, till we have some sort of api to get these limits. Otherwise if all my 200 nodes in ASG gets restarted at the same time will cause us the same problem.

Reply:
Here are Pros and Cons between **Using the hardcoded map** and **Dynamically discover maximum ENIs per instance from AttachmentLimitExceeded**:
## Dynamically discover maximum ENI per instance
### Pros:
* No code change nor new image whenever there is NEW instance type introduced or there any improvements on EC2 instance ENI limit.
### Cons:
* there is ONE extra call to discover the EC2 instance ENI limit during the life time of L-IPAMD.

## Using the hardcode map
### Pros:
* no extra ONE call to discovery the EC2 instance ENI limit
### Cons:
* Whenever there is a New instance type and there is any enhancement on EC2 instance ENI limits, we need to have a code change and new image
* This can add operation cost that you need to upgrade the aws-vpc-cni-k8s daemon set frequently to support new instance type and new EC2 ENI enhancement

Reply:
Agree. But we can also populate hardcoded map for known instance type and for unknown ones we can populate dynamically and store it.

Reply:
Also its not just one call, one call for CreateNetworkInterface, one for AttachNetworkInterface and other for DeleteNetworkInterface, which we can avoid using map. My only concern is if its larger cluster with 1000 nodes then all nodes will perform this operation right?

Reply:
If static Map is used,  then it is operation hassle to upgrade aws-vpc-cni-k8s daemon for 1000 nodes every time there is a new instance type or any change in EC2 instance

Reply:
Keep in mind, NOT all 1000 nodes will perform this operation unless its available IP address are below threshold.

Reply:
That's true.